### PR TITLE
Fix missing update of botanist state with dynamic node network CIDR determined by infrastructure provider

### PR DESCRIPTION
**Pull request by MichaelEischer**
**How to categorize this PR?**

/area control-plane,networking
/kind bug

**What this PR does / why we need it**:

When creating shoots using HA control planes and an infrastructure provider using a [dynamic nodes network](https://github.com/gardener/gardener/blob/master/docs/extensions/infrastructure.md#dynamic-nodes-network-for-shoot-clusters), the shoots will initially be stuck while the seed kube API server tries to reach the shoot nodes via the VPN connection. The shoot only becomes healthy once a new reconciliation run is triggered.

With a dynamically determined nodes network, the vpn seed containers are initially created with an empty `NODE_NETWORK` environment variable. This causes missing routes for an HA setup, see https://github.com/gardener/vpn2/blob/b088ef871ede0e77120e0ce1aaf46a5dfadad0df/seed-server/network-connection.sh#L183 . The underlying reason for this bug is that botanist first collects default configurations for all control plane components (e.g., https://github.com/gardener/gardener/blob/ddb8b312e495616a0d0b1f8ecdf835c3b941b9d1/pkg/operation/botanist/botanist.go#L151), later on triggers the infrastructure reconciliation and waits for it to set the node network (https://github.com/gardener/gardener/blob/ddb8b312e495616a0d0b1f8ecdf835c3b941b9d1/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L219) and finally deploys the VPN without updating the initially set node network (https://github.com/gardener/gardener/blob/ddb8b312e495616a0d0b1f8ecdf835c3b941b9d1/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L369).

This PR updates the botanist flow to update the node network right before deploying CoreDNS / the VPN seed containers.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

I'm not sure that I've found all relevant places.

**Release note**:

```bugfix operator
Cluster creation with highly available control planes and an infrastructure extension that uses dynamic node networks is no longer delayed by a failing VPN connection before the first reconciliation.
```

### CODE REVIEW

File: pkg/operation/botanist/vpnseedserver.go [hunk 1]
```diff
@@ -79,8 +79,8 @@ func (b *Botanist) DefaultVPNSeedServer() (vpnseedserver.Interface, error) {
 		Network: vpnseedserver.NetworkValues{
 			PodCIDR:     b.Shoot.Networks.Pods.String(),
 			ServiceCIDR: b.Shoot.Networks.Services.String(),
-			NodeCIDR:    pointer.StringDeref(b.Shoot.GetInfo().Spec.Networking.Nodes, ""),
-			IPFamilies:  b.Shoot.GetInfo().Spec.Networking.IPFamilies,
+			// NodeCIDR is set in DeployVPNServer
```
Comment (ScheererJ) [inline pkg/operation/botanist/vpnseedserver.go]: The comment is just stating the obvious. Please add why this is the case (spoiler: dynamic node network CIDRs).

File: pkg/operation/botanist/vpnseedserver.go [hunk 2]
```diff
@@ -107,6 +107,9 @@ func (b *Botanist) DefaultVPNSeedServer() (vpnseedserver.Interface, error) {
 
 // DeployVPNServer deploys the vpn-seed-server.
 func (b *Botanist) DeployVPNServer(ctx context.Context) error {
+	if network := b.Shoot.GetInfo().Spec.Networking; network != nil {
+		b.Shoot.Components.ControlPlane.VPNSeedServer.SetNodeNetworkCIDR(network.Nodes)
+	}
```
Comment (ScheererJ) [inline pkg/operation/botanist/vpnseedserver.go]: Let's get rid of the `if` block and always execute the setter.

The VPN is only deployed for clusters with workers, which implies `.spec.networking` to be set. Workerless clusters do not have `.spec.networking`, but also do not deploy the VPN.

File: pkg/operation/botanist/kubeapiserver.go [hunk 3]
```diff
@@ -56,14 +55,10 @@ func (b *Botanist) DefaultKubeAPIServer(ctx context.Context) (kubeapiserver.Inte
...
 
-	if b.Shoot.GetInfo().Spec.Networking != nil {
-		nodes = b.Shoot.GetInfo().Spec.Networking.Nodes
-	}
-
 	if !b.Shoot.IsWorkerless {
 		vpnConfig.Enabled = true
 		vpnConfig.PodNetworkCIDR = pods
-		vpnConfig.NodeNetworkCIDR = nodes
+		// NodeNetworkCIDR is set on deployment
```
Comment (ScheererJ) [inline pkg/operation/botanist/kubeapiserver.go]: Please add why this is done.

File: pkg/operation/botanist/coredns.go [hunk 4]
```diff
@@ -43,11 +43,11 @@ func (b *Botanist) DefaultCoreDNS() (coredns.Interface, error) {
...
-		ClusterDomain:                   gardencorev1beta1.DefaultDomain,
-		ClusterIP:                       b.Shoot.Networks.CoreDNS.String(),
-		Image:                           image.String(),
-		PodNetworkCIDR:                  b.Shoot.Networks.Pods.String(),
-		NodeNetworkCIDR:                 b.Shoot.GetInfo().Spec.Networking.Nodes,
+		ClusterDomain:  gardencorev1beta1.DefaultDomain,
+		ClusterIP:      b.Shoot.Networks.CoreDNS.String(),
+		Image:          image.String(),
+		PodNetworkCIDR: b.Shoot.Networks.Pods.String(),
+		// NodeNetworkCIDR is set on deployment
```
Comment (ScheererJ) [inline pkg/operation/botanist/coredns.go]: Again, please explain why this is done like this.

File: pkg/operation/botanist/coredns.go [hunk 5]
```diff
@@ -77,6 +77,9 @@ func (b *Botanist) DeployCoreDNS(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
+	if network := b.Shoot.GetInfo().Spec.Networking; network != nil {
+		b.Shoot.Components.SystemComponents.CoreDNS.SetNodeNetworkCIDR(network.Nodes)
+	}
```
Comment (ScheererJ) [inline pkg/operation/botanist/coredns.go]: `coredns` is only deployed in clusters with workers. Therefore, please get rid of the `if` check for readability.

File: pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go [hunk 6]
```diff
@@ -259,7 +258,7 @@ func (r *Reconciler) runDeleteShootFlow(ctx context.Context, o *operation.Operat
 				waitUntilEtcdReady,
 				waitUntilKubeAPIServerServiceIsReady,
 				waitUntilControlPlaneReady,
-			).InsertIf(!staticNodesCIDR),
+			),
```
Comment (ScheererJ) [inline pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go:261]: Could you please share why you think that this is dead code?

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Thanks a lot for identifying the issue and creating the pull request.

I have a few comments/questions.

Could you please also adapt the pull request description as it is misleading with regards to `vpn-shoot`, which does not play a role here at all (looking at the changes)?

File: pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go (hunk 6 above)
Comment (MichaelEischer) [reply pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go:261]: "dead code" may not be the optimal term, however, `InsertIf(condition bool, iders ...TaskIDer)` https://github.com/gardener/gardener/blob/fc31b86d6da5a30d87a9e62bbd61ff9aaf0c9de0/pkg/utils/flow/taskid.go#L66 with an empty `iders` list does not modify the TaskIDs set. That is, the call to `InsertIf` has no effect and can simply be removed.

File: pkg/operation/botanist/coredns.go (hunk 5 above)
Comment (MichaelEischer) [reply pkg/operation/botanist/coredns.go]: I've removed the if condition.

File: pkg/operation/botanist/vpnseedserver.go (hunk 2 above)
Comment (MichaelEischer) [reply pkg/operation/botanist/vpnseedserver.go]: I've removed the if condition.

File: pkg/operation/botanist/vpnseedserver.go (hunk 1 above)
Comment (MichaelEischer) [reply pkg/operation/botanist/vpnseedserver.go]: I've extended the this comment and the two other ones. Is that what you had in mind?

**Comment by MichaelEischer** [comment]
> Could you please also adapt the pull request description as it is misleading with regards to `vpn-shoot`, which does not play a role here at all (looking at the changes)?

I've updated the description and replaced the `vpn seed/shoot` references with `vpn seed containers`.

The other review comments are also addressed. PTAL.

File: pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go (hunk 6 above)
Comment (ScheererJ) [reply pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go:261]: Good catch.

Indeed this code does not seem to cause any effect regardless of the condition. Hence, removing it after over three years seems like a good choice.

**Comment by ScheererJ** [comment]
The change itself looks good now. Please fix the pull request description, though.

**Comment by MichaelEischer** [comment]
> , which points to `b.DefaultVPNShoot()`.

Oh, sorry, I've totally missed that I've referenced the wrong line. Now, it points to `b.DefaultVPNSeedServer() ` as intended.

**Comment by ScheererJ** [comment]
Thanks for the changes.

/lgtm

**Pull request merged**